### PR TITLE
KAFKA-12331: Use LEO for the base offset of LeaderChangeMessage batch

### PR DESCRIPTION
**Pull request by jsancio**
The `KafkaMetadataLog` implementation of `ReplicatedLog` validates that batches appended using `appendAsLeader` and `appendAsFollower` have an offset that matches the LEO. This is enforced by `KafkaRaftClient` and `BatchAccumulator`. When creating control batches for the `LeaderChangeMessage` the default base offset of `0` was being used instead of using the LEO. This is fixed by:

1. Changing the implementation for `MockLog` to validate against this and throw an `RuntimeException` if this invariant is violated.
2. Always create a batch for `LeaderChangeMessage` with an offset equal to the LEO.

*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

File: raft/src/test/java/org/apache/kafka/raft/MockLog.java [hunk 1]
```diff
@@ -250,30 +249,7 @@ private LogEntry buildEntry(Long offset, SimpleRecord record) {
...
-        AtomicLong offsetSupplier = new AtomicLong(baseOffset);
-        for (RecordBatch batch : records.batches()) {
-            List<LogEntry> entries = buildEntries(batch, record -> offsetSupplier.getAndIncrement());
-            appendBatch(new LogBatch(epoch, batch.isControlBatch(), entries));
-        }
-
-        return new LogAppendInfo(baseOffset, offsetSupplier.get() - 1);
-    }
-
-    LogAppendInfo appendAsLeader(Collection<SimpleRecord> records, int epoch) {
```
Comment (jsancio) [inline raft/src/test/java/org/apache/kafka/raft/MockLog.java:266]: This method was only used for tests. To simplify the implementation, I removed this method and added a helper method to `MockLogTest`.

**Comment by jsancio** [comment]
@hachikuji this PR is ready for review. Thanks!

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java [hunk 2]
```diff
@@ -106,7 +106,8 @@ public void testFetchRequestWithLargerLastFetchedEpoch() throws Exception {
         OffsetAndEpoch oldestSnapshotId = new OffsetAndEpoch(3, 2);
 
         RaftClientTestContext context = new RaftClientTestContext.Builder(localId, voters)
-            .appendToLog(oldestSnapshotId.offset, oldestSnapshotId.epoch, Arrays.asList("a", "b", "c"))
+            .appendToLog(oldestSnapshotId.epoch, Arrays.asList("a", "b", "c"))
+            .appendToLog(oldestSnapshotId.epoch, Arrays.asList("a", "b", "c"))
```
Comment (hachikuji) [inline raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java]: nit: could we use a different records from the first append in these tests? It make them a little easier to follow

**Review by hachikuji — APPROVED** [review]
LGTM. Just a minor comment on the tests.

File: raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java (hunk 2 above)
Comment (jsancio) [reply raft/src/test/java/org/apache/kafka/raft/KafkaRaftClientSnapshotTest.java]: Yeah. The tests are getting hard to follow and write. When we have finish some of the more important tasks, I want to revisit `RaftClientTestContext` API.

**Comment by jsancio** [comment]
Ran the following commands locally:
```
$ ./gradlew -version

------------------------------------------------------------
Gradle 6.8.1
------------------------------------------------------------

Build time:   2021-01-22 13:20:08 UTC
Revision:     31f14a87d93945024ab7a78de84102a3400fa5b2

Kotlin:       1.4.20
Groovy:       2.5.12
Ant:          Apache Ant(TM) version 1.10.9 compiled on September 27 2020
JVM:          1.8.0_282 (Private Build 25.282-b08)
OS:           Linux 5.8.0-7642-generic amd64

./gradlew -PscalaVersion=2.12 clean compileJava compileScala compileTestJava compileTestScala spotlessScalaCheck checkstyleMain checkstyleTest spotbugsMain rat --profile --no-daemon --continue -PxmlSpotBugsReport=true

./gradlew -PscalaVersion=2.13 clean compileJava compileScala compileTestJava compileTestScala spotlessScalaCheck checkstyleMain checkstyleTest spotbugsMain rat --profile --no-daemon --continue -PxmlSpotBugsReport=true

./gradlew -PscalaVersion=2.12 unitTest integrationTest --profile --no-daemon --continue -PtestLoggingEvents=started,passed,skipped,failed -PignoreFailures=true -PmaxParallelForks=2 -PmaxTestRetries=1 -PmaxTestRetryFailures=5
```

**Comment by jsancio** [comment]
We got the following failures:
```
core:
    ConfigCommandTest. shouldFailIfUnresolvableHost()
    ConfigCommandTest. shouldFailIfUnresolvableHostUsingZookeeper()
    DelegationTokenCommandTest. testDelegationTokenRequests()
    MetricsTest. testMetrics()
    DynamicConfigChangeTest. testIpHandlerUnresolvableAddress()
    DynamicConfigTest. shouldFailIpConfigsWithBadHost()

```
